hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientTransactionManagerServiceImpl.java

```diff
@@ -106,23 +108,50 @@ public String getGroupName() {
     }
 
     public ClientConnection connect() throws Exception {
-        Exception lastError = null;
-        int count = 0;
-        while (count < RETRY_COUNT) {
+        long startTimeMillis = System.currentTimeMillis();
+        AbstractClientInvocationService invocationService = (AbstractClientInvocationService) client.getInvocationService();
+        ClientConfig clientConfig = client.getClientConfig();
+        boolean smartRouting = clientConfig.getNetworkConfig().isSmartRouting();
+
+        Exception lastException = null;
+        while (System.currentTimeMillis() - startTimeMillis < invocationService.getInvocationTimeoutMillis()) {
+            if (!client.getLifecycleService().isRunning()) {
+                throw new HazelcastClientNotActiveException("Client is shutdown");
+            }
             try {
-                final Address randomAddress = getRandomAddress();
-                return (ClientConnection) client.getConnectionManager().getOrConnect(randomAddress);
+                if (smartRouting) {
+                    return tryConnectSmart();
+                } else {
+                    return tryConnectUnisocket();
+                }
             } catch (Exception e) {
-                lastError = e;
+                lastException = e;
+                if (e instanceof HazelcastClientOfflineException) {
+                    throw e;
+                }
             }
-            count++;
+            Thread.sleep(invocationService.getInvocationRetryPauseMillis());
```

Comment:
This causes an extra sleep when throwing an exception, you can use "do while" to eliminate the extra unneeded sleep.